libpromises/eval_context.c

```diff
@@ -3043,3 +3081,48 @@ void EvalContextAllClassesLoggingEnable(EvalContext *ctx, bool enable)
     assert (ctx);
     Nova_ClassHistoryEnable(&(ctx->all_classes), enable);
 }
+
+void EvalContextPushBundleName(const EvalContext *ctx, const char *bundle_name)
+{
+    assert (ctx);
+    StringSetAdd(ctx->bundle_names, xstrdup(bundle_name));
+}
+
+const StringSet *EvalContextGetBundleNames(const EvalContext *ctx)
+{
+    assert (ctx);
+    return ctx->bundle_names;
+}
+
+void EvalContextPushRemoteVarPromise(EvalContext *ctx, const char *bundle_name, const Promise *pp)
+{
+    assert (ctx);
+
+    /* initiliaze the map if needed */
+    if (ctx->remote_var_promises == NULL)
+    {
+        ctx->remote_var_promises = RemoteVarPromisesMapNew();
+    }
+
+    Seq *promises = RemoteVarPromisesMapGet(ctx->remote_var_promises, bundle_name);
+    if (promises == NULL)
+    {
+        /* initialize the sequence if needed */
+        /* ItemDestroy == NULL because we need to store the exact pointers not
+         * copies */
+        promises = SeqNew(10, NULL);
+        RemoteVarPromisesMapInsert(ctx->remote_var_promises, xstrdup(bundle_name), promises);
+    }
+    /* intentionally not making a copy here, we need the exact pointer */
+    SeqAppend(promises, (void *) pp);
```

Comment:
Why take a const pointer and discard the const?

Reply:
Because we don't have a *const Sequence*. Taking ``const Promise *`` is suggesting that it won't be modified and/or deallocated. But we don't have a sequence type that would provide such guarantees and expressed them in its type signatures. So we have to rely on the implementation when ``ItemDestroy == NULL``. And we want to avoid compiler warnings.